frameworks/Java/armeria/src/main/java/hello/services/PostgresFortunesService.java

```diff
@@ -44,43 +48,51 @@ public PostgresFortunesService() {
   }
 
   @Get("/fortunes")
-  public HttpResponse fortunes() throws SQLException, IOException {
-    List<Fortune> fortunes = getFortunes();
-    fortunes.add(
-        new Fortune(0, "Additional fortune added at request time."));
-    fortunes.sort(Comparator.comparing(Fortune::getMessage));
-
-    return HttpResponse.of(
-        HttpHeadersHelper.getHttpHeader(MediaType.HTML_UTF_8),
-        HttpData.ofUtf8(buildMustacheTemplate(fortunes))
-        );
+  public CompletableFuture<HttpResponse> fortunes() {
+    return getFortunes().thenApply(fortunes -> {
+      fortunes.add(
+          new Fortune(0, "Additional fortune added at request time."));
+      fortunes.sort(Comparator.comparing(Fortune::getMessage));
+
+      return HttpResponse.of(
+              HttpStatus.OK, MediaType.HTML_UTF_8, buildMustacheTemplate(fortunes));
+    });
   }
 
-  private List<Fortune> getFortunes() throws SQLException {
-    List<Fortune> fortunes = new ArrayList<>();
-
-    try (final Connection connection = dataSource.getConnection();
-         final PreparedStatement statement =
-             connection.prepareStatement(SELECT_QUERY);
-         final ResultSet resultSet = statement.executeQuery()) {
-
-      while (resultSet.next()) {
-        fortunes.add(
-            new Fortune(
-                resultSet.getInt(1),
-                resultSet.getString(2)));
-      }
-    }
-    return fortunes;
+  private CompletableFuture<List<Fortune>> getFortunes() {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          List<Fortune> fortunes = new ArrayList<>();
+
+          try (final Connection connection = dataSource.getConnection();
+               final PreparedStatement statement =
+                   connection.prepareStatement(SELECT_QUERY);
+               final ResultSet resultSet = statement.executeQuery()) {
+
+            while (resultSet.next()) {
+              fortunes.add(
+                  new Fortune(
+                      resultSet.getInt(1),
+                      resultSet.getString(2)));
+            }
+          } catch (SQLException e) {
+            throw new IllegalStateException("Database error", e);
+          }
+          return fortunes;
+        }, ServiceRequestContext.current().blockingTaskExecutor()
+    );
   }
 
-  private String buildMustacheTemplate(List<Fortune> fortunes)
-      throws IOException {
+  private byte[] buildMustacheTemplate(List<Fortune> fortunes) {
     Mustache mustache = mustacheFactory.compile("fortunes.mustache");
```

Comment:
Shouldn't we compile the template only once?

Reply:
Mustache apparently automatically caches so this should be ok - it was recommended in the initial submission. https://github.com/TechEmpower/FrameworkBenchmarks/pull/4041#discussion_r217892498

Reply:
Ah OK! Thanks for education.